packages/flutter/lib/src/material/time_picker.dart

```diff
@@ -34,6 +35,359 @@ enum DayPeriod {
   pm,
 }
 
+/// The list of known time of day formats.
+///
+/// This time picker has configurations optimized for these formats. Any other
+/// formats will default to [h_colon_mm_space_a].
+abstract class TimeOfDayFormats {
+  /// The default en_US format.
+  ///
+  /// Used when the configuration for the currently specified
+  /// [MaterialLocalizations.timeOfDayFormat] is not available.
+  static const String h_colon_mm_space_a = 'h:mm a';
+
+  // Reusable patterns.
+  static const String HH_colon_mm = 'HH:mm';
+  static const String H_colon_mm = 'H:mm';
+  static const String HH_dot_mm = 'HH.mm';
+  static const String a_space_h_colon_mm = 'a h:mm';
+
+  // Locale-specific.
+  static const String frenchCanadian = "HH 'h' mm";
+  static const String thai = 'HH:mm น.';
+
+  /// Enumerates formats supported by the material time picker.
+  static const List<String> supportedFormats = const <String>[
+    h_colon_mm_space_a,
+    HH_colon_mm,
+    H_colon_mm,
+    HH_dot_mm,
+    a_space_h_colon_mm,
+    frenchCanadian,
+  ];
+}
+
+class _TimePickerHeaderFormat {
+  const _TimePickerHeaderFormat(this.centrepieceIndex, this.pieces);
+
+  /// Pieces that constitute a time picker header.
+  final List<_TimePickerHeaderPiece> pieces;
+
+  /// Index into the [pieces] list pointing at the piece that contains the
+  /// pivot fragment.
+  final int centrepieceIndex;
+}
+
+/// An unbreakable part of the time picker header.
+///
+/// When the picker is laid out vertically, [fragments] of the piece are laid
+/// out on the same line, with each piece getting its own line.
+class _TimePickerHeaderPiece {
+  const _TimePickerHeaderPiece(this.pivotIndex, this.fragments, { this.bottomMargin: 0.0 });
+
+  /// Fragments this piece is made of.
+  final List<_TimePickerHeaderFragment> fragments;
+
+  /// Index into the [fragments] list, pointing at the fragment that's centered
+  /// horizontally.
+  final int pivotIndex;
+
+  /// Vertical distance between this piece and the next piece.
+  ///
+  /// This property applies only when the header is laid out vertically.
+  final double bottomMargin;
+}
+
+_TimePickerHourFormat _getHourFormat(String formatString) {
+  switch(formatString) {
+    case TimeOfDayFormats.HH_dot_mm:
+    case TimeOfDayFormats.HH_colon_mm:
+    case TimeOfDayFormats.frenchCanadian:
+      return _TimePickerHourFormat.HH;
+    case TimeOfDayFormats.H_colon_mm:
+      return _TimePickerHourFormat.H;
+    default:
+      return _TimePickerHourFormat.h;
```

Comment:
this seems like a bad default. If we really don't recognize the value, we should use `HH`.